### ISSUE TITLE
Radial bar components shifts the filled part relative to the unfilled background

### ISSUE THREAD

**Issue by AigiulGasymova** [recharts/recharts#4264]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/sandbox/recharts-issue-template-forked-tkfvkn?file=%2Fsrc%2Findex.js%3A97%2C24&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltd12cbg0006356k0bwsgbv5%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltd12cbg0002356kv52iomlu%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltd12cbg0003356kizzhjsku%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltd12cbg0005356kwny3r3ur%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltd12cbg0002356kv52iomlu%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltd12cbg0001356kol046c3n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A97%252C%2522startColumn%2522%253A24%252C%2522endLineNumber%2522%253A97%252C%2522endColumn%2522%253A24%257D%255D%257D%255D%252C%2522id%2522%253A%2522cltd12cbg0002356kv52iomlu%2522%252C%2522activeTabId%2522%253A%2522cltd12cbg0001356kol046c3n%2522%257D%252C%2522cltd12cbg0005356kwny3r3ur%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltd12cbg0004356kly3a5oti%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltd12cbg0005356kwny3r3ur%2522%252C%2522activeTabId%2522%253A%2522cltd12cbg0004356kly3a5oti%2522%257D%252C%2522cltd12cbg0003356kizzhjsku%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cltd12cbg0003356kizzhjsku%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce
Created a semicircle chart that fills with color clockwise depending on the current value and has a background, used RadialBarChart with PolarAngleAxis and RadialBar. 

### What is expected?
There is no offset in RadialBar like it was in elder versions - for example, 1.8.2.

### What is actually happening?
When there is a background in RadialBar, the filled part has an outline from the background, and the unfilled part has an outline from the fill color, i.e. the filled part seems to be slightly offset relative to the base with the unfilled part.

| Environment | Info |
|---|---|
| Recharts | v2.12.2 |
| React | 17.0.2 |
| System | Windows 11 Home edition |
| Browser | Chrome 122.0.6261.95 |

**Comment by PavelVanecek** [comment]
If you allow I made screenshots of what it looks like in `recharts: 1.8.2`:
<img width="305" alt="image" src="https://github.com/recharts/recharts/assets/1100170/84f2b349-6106-44c5-a19a-c6861cc90eda">

And this is what it looks like in `recharts: 2.12.2`:
<img width="306" alt="image" src="https://github.com/recharts/recharts/assets/1100170/d797c099-d98e-4593-8a11-d84a9a2a5d5e">

**Comment by PavelVanecek** [comment]
It appears this is a change from 2.9 release (the one where I made my first contribution 😬 )

**Comment by AigiulGasymova** [comment]
@PavelVanecek thanks for the screenshots! Yes, it appears that the issue started reproducing somewhere in the latest versions. Do we have some fixes or workarounds for this?

**Comment by PavelVanecek** [comment]
Looks to me like something that we should fix directly in recharts. Not sure what causes it yet. It looks to me like the problem is in the first X coordinate of `path.d`. Here in my codesandbox it is `65.3` in the foreground RadialBar but `64.5` in the background RadialBar (look for HTML attribute that goes `d="M 64.5..."`:

<img width="1050" alt="image" src="https://github.com/recharts/recharts/assets/1100170/abcb31cc-edf6-45f8-99b6-4dc0fb2f47e7">

If I edit it manually to be the same 64.5 then it renders fine:

<img width="1071" alt="image" src="https://github.com/recharts/recharts/assets/1100170/ada511b1-9cc1-42f2-8e9c-8ca95fbc9764">

Although when I open the same example in 2.8.0 (the working version) it appears that the `65.3000000001` was the correct one:

<img width="1040" alt="image" src="https://github.com/recharts/recharts/assets/1100170/11e2dd13-27ae-49bd-ba01-2c9037fde2ca">

I still don't know which change in 2.9.0 caused this but hopefully this will help narrow it down

**Comment by PavelVanecek** [comment]
Thank you for the details!

**Comment by PavelVanecek** [comment]
I can reproduce in this unit test:

```

  test('renders background in the exact same position as foreground', () => {
    // Reproduction of https://github.com/recharts/recharts/issues/4264
    const preparedData = [{ value: 42, fill: '#241084' }];
    const emptyBackgroundColor = '#D8BDF3';

    const { container, debug } = render(
      <RadialBarChart data={preparedData} height={280} width={340} cx="47%" startAngle={180} endAngle={0}>
        <RadialBar
          background={{
            fill: emptyBackgroundColor,
          }}
          dataKey="value"
          isAnimationActive={false}
        />
      </RadialBarChart>,
    );
    debug();
    const background = container.querySelector('.recharts-radial-bar-background-sector');
    expect(background).toBeInTheDocument();
    const foreground = container.querySelector('.recharts-radial-bar-sector');
    expect(foreground).toBeInTheDocument();
    expect(foreground.getAttribute('d')).toEqual(background.getAttribute('d'));
  });
```

**Comment by ckifer** [comment]
Thanks for helping here @PavelVanecek! 

So maybe we started rounding at some point? I think I remember something like that..

**Comment by ckifer** [comment]
The [calculation](https://github.com/recharts/recharts/blob/3.x/src/util/PolarUtils.ts#L12-L15) itself - this logic hasn't changed n 9 years.

[Path calculation for sector](https://github.com/recharts/recharts/blame/3.x/src/shape/Sector.tsx#L59) - also hasn't changed.

Most likely this PR https://github.com/recharts/recharts/pull/3803/files#diff-445315e1448c74882c065c4b50377b6231aadafa6b72dd3632b233e404c71e45R280 and specifically where we start parsing things as ints

**Comment by AigiulGasymova** [comment]
@ckifer sorry to bother you, but is there any update on this issue? This issue is not urgent and does not have a significant impact on anything other than appearance, and we are considering using the fractional width workaround, but if a fix is coming soon, we'd rather wait for it.

**Comment by AigiulGasymova** [comment]
> Hey @AigiulGasymova
> 
> No update really, the last I looked for this was my last comment above.
> 
> I can't promise anything with when, it really depends on how much free time I have which is pretty random.

Sure, thanks for the answer (and for pretty cool and useful library!). We'll use the workaround and keep tracking this thread just in case!

**Comment by ckifer** [comment]
PR https://github.com/recharts/recharts/pull/4295

**Comment by ckifer** [comment]
released in 2.12.3
